server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesPhase.java

```diff
@@ -48,24 +48,15 @@
 public final class FetchDocValuesPhase implements FetchSubPhase {
 
     @Override
-    public FetchSubPhaseProcessor getProcessor(SearchContext context) throws IOException {
-        if (context.collapse() != null) {
```

Comment:
This logic is moved into FetchContext#docValuesContext(), meaning that we don't need to expose `collapse()` on FetchContext

Reply:
To me it seems off to have non-trivial logic in a simple getter. It feels okay to keep this in `FetchDocValuesPhase` and expose `FetchContext#collapse()` since it has some relevance to the fetch phase ?

Reply:
Would it feel better if the logic was in `FetchContext`'s ctor? I do like the idea of moving separating "which doc values do we need" from "how do we load doc values". We do the same thing with source visiting right now so it kind of makes sense.

Reply:
I disagree, I think.  `FetchContext#collapse()` doesn't really have any relevance to this phase - all the phase cares about is getting a set of fields to fetch from docvalues, and doesn't need to know where the list is from.  And `collapse()` isn't needed for any of the other subphases either.  Whereas inside the FetchContext we have access to the search context as a whole, so we can combine the user request with information from other sources (like the collapse context).  Other methods on FetchContext also combine information from the search context in more or less complex ways (eg `version()` or `fetchScores()`).

Reply:
I don't really like the idea of adding it in the constructor either as it's adding extra state where I don't think we need it?

Reply:
The approach makes sense to me now, it was helpful to talk it through!